File: mesagisto-client/jvm/src/main/kotlin/org/mesagisto/client/Res.kt
@file:Suppress("MemberVisibilityCanBePrivate", "ktlint:standard:no-wildcard-imports")

package org.mesagisto.client

import kotlinx.coroutines.*
import org.mesagisto.client.data.Event
import org.mesagisto.client.data.Packet
import java.nio.file.Path
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.HashSet
import kotlin.coroutines.CoroutineContext
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine
import kotlin.io.path.*

object Res : CoroutineScope {
  val Directory: Path =
    Path(System.getProperty("java.io.tmpdir"))
      .resolve("mesagisto")
      .apply { createDirectories() }

  private val handlers = ConcurrentHashMap<String, HashSet<(Path) -> Unit>>()
  override val coroutineContext: CoroutineContext
    get() = Dispatchers.Default

  init {
    poll()
  }

  private fun poll() =
    launch {
      while (true) {
        delay(200)
        val forRemove = arrayListOf<String>()
        handlers.forEach {
          if (path(it.key).exists()) {
            forRemove.add(it.key)
          }
        }
        for (remove in forRemove) {
          val handlers = handlers.remove(remove) ?: break
          for (handler in handlers) {
            launch {
              handler.invoke(path(remove))
            }
          }
        }
      }
    }

  fun path(name: String): Path = Directory.resolve(name)

  fun tmpPath(name: String): Path = Directory.resolve("$name.tmp")

  fun waitFor(
    name: String,
    handler: (Path) -> Unit,
  ) {
    handlers.getOrPut(name) { HashSet() }.add(handler)
  }

  fun storePhotoId(
    uid: ByteArray,
    fileId: ByteArray = ByteArray(0),
  ) {
    Db.putImageId(uid, fileId)
  }

  fun get(name: String): Path? {
    val file = path(name)
    return if (file.exists()) file else null
  }

  suspend fun file(
    id: ByteArray,
    url: String?,
    room: UUID,
    server: String,
  ): Result<Path> {
    return if (url == null) {
      fileById(id, room, server)
    } else {
      fileByUrl(id, url)
    }
  }

  suspend fun fileById(
    id: ByteArray,
    room: UUID,
    server: String,
  ): Result<Path> =
    runCatching call@{
      val idStr = Base64.encodeToString(id)
      Logger.debug { "通过ID${idStr}缓存文件中" }
      val path = Res.path(idStr)
      if (path.exists()) {
        Logger.trace { "文件存在,返回其路径" }
        return@call path
      }
      val tmpPath = Res.tmpPath(idStr)

      if (tmpPath.exists()) {
        return@call suspendCoroutine { res ->
          Logger.trace { "缓存文件存在,正在等待其下载完毕" }
          waitFor(idStr) { res.resume(it) }
        }
      }

      Logger.trace { "缓存文件不存在,正在请求其URL" }
      val pkt = Packet.new(room, Event.RequestImage(id))
      val resp = Server.request(pkt, server).getOrThrow()

      return@call when (resp) {
        is Event.RespondImage -> {
          fileByUrl(resp.id, resp.url).getOrThrow()
        }
        else -> {
          error("错误的响应")
        }
      }
    }

  suspend fun fileByUrl(
    id: ByteArray,
    url: String,
  ): Result<Path> =
    runCatching call@{
      val idStr = Base64.encodeToString(id)
      Logger.debug { "通过URL${url}缓存文件." }
      val path = path(idStr)
      if (path.exists()) {
        Logger.trace { "文件存在,返回其路径" }
        return@call path
      }
      val tmpPath = tmpPath(idStr)
      if (tmpPath.exists()) {
        Logger.trace { "缓存文件存在,正在等待其下载完毕" }
        suspendCoroutine { res ->
          waitFor(idStr) { res.resume(it) }
        }
      } else {
        Logger.trace { "缓存文件不存在,尝试下载图片" }
        Net.downloadFile(url, tmpPath)
        Logger.trace { "成功下载图片" }
        put(idStr, tmpPath)
        path
      }
    }

  fun put(
    name: String,
    file: Path,
  ): Result<Path> =
    runCatching {
      val path = path(name)
      file.moveTo(path)
      Logger.trace { "将缓存文件 $file 移动至 $path ..." }
      path
    }
}

File: mesagisto-client/rust/src/res.rs
use std::{panic, path::PathBuf, time::Duration};

use arcstr::ArcStr;
use color_eyre::eyre::Result;
use dashmap::DashMap;
use lateinit::LateInit;
use sled::IVec;
use tokio::{sync::oneshot, task::JoinHandle, time::timeout};
use tracing::trace;
use uuid::Uuid;

use crate::{
  data::{events::Event, Packet},
  db::DB,
  net::NET,
  server::SERVER,
  ResultExt,
};

#[derive(Singleton, Default)]
pub struct Res {
  pub directory: LateInit<PathBuf>,
  pub handlers: DashMap<ArcStr, Vec<oneshot::Sender<PathBuf>>>,
  handle: LateInit<JoinHandle<()>>,
}
impl Res {
  pub async fn init(&self) {
    let path = {
      let mut dir = std::env::temp_dir();
      dir.push("mesagisto");
      dir
    };
    tokio::fs::create_dir_all(path.as_path()).await.unwrap();
    self.directory.init(path);
    RES.poll().await;
  }

  pub fn get(&self, name: &ArcStr) -> Option<PathBuf> {
    let path = self.path(name);
    if path.exists() {
      Some(path)
    } else {
      None
    }
  }

  async fn poll(&self) {
    let handle: JoinHandle<_> = tokio::spawn(async {
      let mut interval = tokio::time::interval(Duration::from_millis(200));
      loop {
        let mut for_remove = vec![];
        for entry in &RES.handlers {
          let path = RES.path(entry.key());
          if path.exists() {
            for_remove.push((entry.key().to_owned(), path));
          }
        }
        for_remove.into_iter().for_each(|v| {
          if let Some((.., handler_list)) = RES.handlers.remove(&v.0) {
            for handler in handler_list {
              handler.send(v.1.to_owned()).log();
            }
          }
        });
        interval.tick().await;
      }
    });
    self.handle.init(handle);
  }

  pub fn path(&self, id: &ArcStr) -> PathBuf {
    let mut path = self.directory.clone();
    path.push(id.as_str());
    path
  }

  pub fn tmp_path(&self, id: &ArcStr) -> PathBuf {
    let mut path = self.directory.clone();
    path.push(format!("{}.tmp", id));
    path
  }

  pub async fn wait_for(&self, id: &ArcStr) -> Result<PathBuf> {
    let (sender, receiver) = oneshot::channel();
    self.handlers.entry(id.clone()).or_default().push(sender);
    let path = tokio::time::timeout(Duration::from_secs_f32(13.0), receiver).await??;
    Ok(path)
  }

  pub fn put_image_id<U, F>(&self, uid: U, file_id: F)
  where
    U: AsRef<[u8]>,
    F: Into<IVec>,
  {
    DB.put_image_id(uid, file_id);
  }

  pub async fn file(
    &self,
    id: &Vec<u8>,
    url: &Option<ArcStr>,
    room: &Uuid,
    server: &ArcStr,
  ) -> Result<PathBuf> {
    match url {
      Some(url) => self.file_by_url(id, url).await,
      None => self.file_by_uid(id, room, server).await,
    }
  }

  pub async fn file_by_uid(&self, uid: &Vec<u8>, room: &Uuid, server: &ArcStr) -> Result<PathBuf> {
    use crate::data::Payload;
    let uid_str: ArcStr = base64_url::encode(uid).into();
    trace!("Caching file by uid {}", uid_str);
    let path = RES.path(&uid_str);
    if path.exists() {
      trace!("File exists,return the path");
      return Ok(path);
    }
    let tmp_path = RES.tmp_path(&uid_str);
    if tmp_path.exists() {
      trace!("TmpFile exists,waiting for the file downloading");
      return RES.wait_for(&uid_str).await;
    }
    trace!("TmpFile dont exist,requesting image url");
    let event: Event = Event::RequestImage { id: uid.clone() };
    // fixme error handling
    let packet = Packet::new(room.to_owned(), event.into())?;
    // fixme timeout check
    let packet = timeout(Duration::from_secs(7), SERVER.request(packet, server)).await??;

    match packet.decrypt()? {
      Payload::EventPayload(Event::RespondImage { id, url }) => self.file_by_url(&id, &url).await,
      _ => panic!("Not correct response"),
    }
  }

  pub async fn file_by_url(&self, id: &Vec<u8>, url: &ArcStr) -> Result<PathBuf> {
    let id_str: ArcStr = base64_url::encode(id).into();
    let path = RES.path(&id_str);
    if path.exists() {
      return Ok(path);
    }

    let tmp_path = RES.tmp_path(&id_str);
    if tmp_path.exists() {
      Ok(RES.wait_for(&id_str).await?)
    } else {
      // fixme error handling
      NET.download(url, &tmp_path).await?;
      tokio::fs::rename(&tmp_path, &path).await?;
      Ok(path)
    }
  }

  pub async fn put_file(&self, id: &Vec<u8>, file: &PathBuf) -> Result<PathBuf> {
    let id_str: ArcStr = base64_url::encode(id).into();
    let path = RES.path(&id_str);
    tokio::fs::rename(&file, &path).await?;
    Ok(path)
  }
}
